src/m3ninx/index/regexp_test.go

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"fmt"
+	"regexp/syntax"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestEnsureSyntaxPerlTreatsAnchorsAsTextTerminator(t *testing.T) {
+	// Test to ensure future compatibility with changes in `regexp/syntax`.
+	//
+	// We require that '^' and '$' only match input terminating characters (i.e.
+	// text boundaries, not line boundaries within the input). The line of code
+	// below ensures that syntax.Perl does the same.
+	require.NotZero(t, syntax.Perl&syntax.OneLine)
+
+	// ensure our `parseRegexp` internal function uses the right flags too.
+	re, err := parseRegexp(".*")
+	require.NoError(t, err)
+	require.NotZero(t, re.Flags&syntax.OneLine)
+}
+
+func TestEnsureRegexpUnachored(t *testing.T) {
+	testCases := []testCase{
+		testCase{
+			name:           "naked ^",
+			input:          "^",
+			expectedOutput: "(?:)",
+		},
+		testCase{
+			name:           "naked $",
+			input:          "$",
+			expectedOutput: "(?:)",
+		},
+		testCase{
+			name:           "empty string ^$",
+			input:          "^$",
+			expectedOutput: "",
+		},
+		testCase{
+			name:           "invalid naked concat ^$",
+			input:          "$^",
+			expectedOutput: "(?-m:$)\\A",
+		},
+		testCase{
+			name:           "simple case of ^",
+			input:          "^abc",
+			expectedOutput: "abc",
+		},
+		testCase{
+			name:           "simple case of $",
+			input:          "abc$",
+			expectedOutput: "abc",
+		},
+		testCase{
+			name:           "simple case of both ^ & $",
+			input:          "^abc$",
+			expectedOutput: "abc",
+		},
+		testCase{
+			name:           "weird case of internal ^",
+			input:          "^a^bc$",
+			expectedOutput: "a\\Abc",
```

Comment:
Hm, what's `"\\A"` out of interest?

Also, wish that they exported `dumpRegexp` from stdlib so you could use that to test... you could copy paste if you want to use that kind of output to test against (e.g. `"cat{lit{foo},lit{bar}}"`):
https://golang.org/src/regexp/syntax/parse_test.go#L315

Reply:
re: `dumpRegexp()` solid idea, will do.

Reply:
done